pkg/services/ngalert/state/state_tracker.go

```diff
@@ -24,77 +31,132 @@ type cache struct {
-		newState := st.getNextState(uid, result)
-		if newState != currentState.State {
-			currentState.State = newState
-			changedStates = append(changedStates, currentState)
+		if s, ok := st.setNextState(uid, result); ok {
+			changedStates = append(changedStates, s)
 		}
-		st.stateCache.update(currentState)
 	}
+	st.Log.Debug("returning changed states to scheduler", "count", len(changedStates))
 	return changedStates
 }
 
-func (st *StateTracker) getNextState(uid string, result eval.Result) eval.State {
-	currentState := st.stateCache.getOrCreate(uid, result)
-	if currentState.State == result.State {
-		return currentState.State
-	}
-
+//Set the current state based on evaluation results
+//return the state and a bool indicating whether a state transition occurred
+func (st *StateTracker) setNextState(uid string, result eval.Result) (AlertState, bool) {
+	currentState := st.getOrCreate(uid, result)
+	st.Log.Debug("setting alert state", "uid", uid)
 	switch {
 	case currentState.State == result.State:
-		return currentState.State
+		st.Log.Debug("no state transition", "cacheId", currentState.CacheId, "state", currentState.State.String())
+		currentState.EvaluatedAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.Results = append(currentState.Results, result.State)
+		st.set(currentState)
+		return currentState, false
 	case currentState.State == eval.Normal && result.State == eval.Alerting:
-		return eval.Alerting
+		st.Log.Debug("state transition from normal to alerting", "cacheId", currentState.CacheId)
+		currentState.State = eval.Alerting
+		currentState.EvaluatedAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.StartsAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.EndsAt = strfmt.DateTime(result.EvaluatedAt.Add(40 * time.Second))
+		currentState.Results = append(currentState.Results, result.State)
+		st.set(currentState)
+		return currentState, true
 	case currentState.State == eval.Alerting && result.State == eval.Normal:
-		return eval.Normal
+		st.Log.Debug("state transition from alerting to normal", "cacheId", currentState.CacheId)
+		currentState.State = eval.Normal
+		currentState.EvaluatedAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.EndsAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.Results = append(currentState.Results, result.State)
+		st.set(currentState)
+		return currentState, true
 	default:
-		return eval.Alerting
+		return currentState, false
+	}
+}
+
+func (st *StateTracker) cleanUp() {
+	ticker := time.NewTicker(time.Duration(60) * time.Minute)
+	st.Log.Debug("starting cleanup process", "intervalMinutes", 60)
+	wg := sync.WaitGroup{}
+	for {
+		select {
+		case <-ticker.C:
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				st.trim()
+			}()
+			wg.Wait()
+		case <-st.quit:
+			st.Log.Debug("stopping cleanup process", "now", time.Now())
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+func (st *StateTracker) trim() {
+	st.Log.Info("trimming alert state cache", "now", time.Now())
```

Comment:
I'll double check. I was following a pattern that I saw in schedule.go https://github.com/grafana/grafana/blob/4a6ad4ad88d856dc6136872ac2d3f4443c399166/pkg/services/ngalert/schedule/schedule.go#L77-L78

Reply:
I see the time in the logs. The "now" that you pointed might be telling you for what time was it evaluating the alert since it's different from time.Now() and coming from ctx.

Reply:
what @codesome says; in the above snippet the `ctx.now` is the tick when the rule was scheduled to get evaluated as opposed to the logging time.